### PR TITLE
Low-Power Device Management Library API Review

### PR DESCRIPTION
**Pull request by ppetrosh**
- The PR is for the initial API proposal only. Before we go ahead with the implementation it would be good if you @hfiennes and @terrencebarr review and provide your comments.
- Please note: the library name is not final and any ideas are very welcome.

### CODE REVIEW

File: LPDeviceManager.device.lib.nut [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+     * The function that implements an action to be performed.
+     *
+     * @callback action-callback
+     * @param {function} done - the function to be called by the action when done.
+     */
+    function doAsyncAndSleep(action, sleepTime, timeout = LP_DEFAULT_TIMEOUT_SEC) {
+        // Schedule deep sleep after timeout
+        imp.wakeup(timeout, function() {
+            sleepFor(sleepTime);
+        });
```
Comment (betzrhodes) [inline LPDeviceManager.device.lib.nut]: missing bindenv

Comment (ppetrosh) [reply LPDeviceManager.device.lib.nut]: Thank you @betzrhodes, good catch. Fixed.

**Comment by betzrhodes** [comment]
@smittytone This is ready for a beta release. Can you review the docs and make any changes needed on develop. I'll publish when you are done. Thanks!

File: README.md [hunk 2]
```diff
@@ -1,75 +1,77 @@
...
+| *onInterrupt* | Callback to be triggered when a wakeup pin is asserted. The callback takes no parameters |
+| *onPowerRestored* | Callback to be triggered when the imp's V<sub>BAT</sub> is powered during a cold start. The callback takes no parameters. **Note** This wake reason is only available on impOS™ 40 and above |
+| *defaultOnWake* | Callback that catches [all other wake reasons](https://developer.electricimp.com/api/hardware/wakereason), and takes the impOS wake reason constant as its argument |
 
 #### Example ####
 
 ```squirrel
 // Wake up on timer flow
 function onScheduledWake() {
+    server.log("Device woke according to schedule");
```
Comment (betzrhodes) [inline README.md]: FYI - With connection manager defaults the device is not connected here, so this log will not show.

File: README.md [hunk 3]
```diff
@@ -143,34 +145,32 @@ Nothing.
...
+        if ("error" in reading) {
             // Log error
-            server.error("Temperature/Humidity reading error: " + res.error);
+            server.error("Temperature/Humidity reading error: " + reading.error);
         } else {
             // Log reading
-            server.log("Temperature: " + res.temperature + "°C Humidity: " + res.humidity + "%");
-            // Go to sleep
-            done();
+            server.log("Temperature: " + reading.temperature + "°C Humidity: " + reading.humidity + "%");
```
Comment (betzrhodes) [inline README.md:151]: This example should call done() parameter.

File: README.md [hunk 4]
```diff
@@ -273,35 +265,27 @@ lpm.onDisconnect(function(expected) {
...
 #### Return Value ####
 
-Boolean, `true` if the device is connected and `false` otherwise.
+Boolean &mdash; `true` if the device is connected, otherwise `false`.
 
 #### Example ####
 
 ```squirrel
-if (lpm.isConnected()) server.log("Device is connected");
+server.log("Device is " + (lpm.isConnected() ? "connected" : "disconnected"));
```
Comment (betzrhodes) [inline README.md]: The device should not log when disconnected. We should not be recommending this practice in examples.

File: README.md (hunk 3 above)
Comment (smittytone) [reply README.md:151]: I took that out because I couldn't see where *done* was being passed into _readTemp()_ from so assumed it was a mistake.

**Review by smittytone — APPROVED** [review]

File: README.md (hunk 3 above)
Comment (betzrhodes) [reply README.md:151]: There was a mistake. doAsyncAndSleep action function takes a parameter, a function that should be triggered when the async action is completed. I've updated the docs and the example.

**Pull request merged**